File: backend-node/src/employees/certificate-storage.service.ts
import { BadRequestException, Injectable, NotFoundException } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { randomUUID } from 'crypto';
import { createReadStream, ReadStream } from 'fs';
import { mkdir, unlink, writeFile } from 'fs/promises';
import { basename, resolve, sep } from 'path';

interface StoredCertificate {
  storedName: string;
  originalName: string;
  mimeType: string;
  size: number;
}

interface DetectedFileType {
  extension: 'pdf' | 'jpg' | 'png';
  mimeType: 'application/pdf' | 'image/jpeg' | 'image/png';
}

@Injectable()
export class CertificateStorageService {
  private readonly uploadDirectory: string;

  constructor(config: ConfigService) {
    this.uploadDirectory = resolve(config.getOrThrow<string>('UPLOAD_DIR'));
  }

  async store(file: Express.Multer.File): Promise<StoredCertificate> {
    const detectedType = this.detectType(file.buffer);
    if (!detectedType) {
      throw new BadRequestException('Certificate must be a valid PDF, JPEG, or PNG file');
    }

    await mkdir(this.uploadDirectory, { recursive: true, mode: 0o750 });
    const storedName = `${randomUUID()}.${detectedType.extension}`;
    await writeFile(this.safePath(storedName), file.buffer, { flag: 'wx', mode: 0o640 });

    return {
      storedName,
      originalName: this.safeOriginalName(file.originalname),
      mimeType: detectedType.mimeType,
      size: file.size,
    };
  }

  open(storedName: string): ReadStream {
    return createReadStream(this.safePath(storedName));
  }

  async remove(storedName: string | null): Promise<void> {
    if (!storedName) return;
    try {
      await unlink(this.safePath(storedName));
    } catch (error: unknown) {
      if ((error as NodeJS.ErrnoException).code !== 'ENOENT') throw error;
    }
  }

  ensurePresent(file: Express.Multer.File | undefined): Express.Multer.File {
    if (!file) throw new BadRequestException('A certificate file is required');
    return file;
  }

  private safePath(storedName: string): string {
    const fullPath = resolve(this.uploadDirectory, storedName);
    if (!fullPath.startsWith(`${this.uploadDirectory}${sep}`)) {
      throw new NotFoundException('Certificate not found');
    }
    return fullPath;
  }

  private safeOriginalName(originalName: string): string {
    return basename(originalName.replaceAll('\0', '')).slice(0, 255) || 'certificate';
  }

  private detectType(buffer: Buffer): DetectedFileType | null {
    if (buffer.length >= 5 && buffer.subarray(0, 5).toString('ascii') === '%PDF-') {
      return { extension: 'pdf', mimeType: 'application/pdf' };
    }
    if (buffer.length >= 3 && buffer[0] === 0xff && buffer[1] === 0xd8 && buffer[2] === 0xff) {
      return { extension: 'jpg', mimeType: 'image/jpeg' };
    }
    if (
      buffer.length >= 8 &&
      buffer.subarray(0, 8).equals(Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]))
    ) {
      return { extension: 'png', mimeType: 'image/png' };
    }
    return null;
  }
}

File: backend-springboot/src/main/java/com/compnet/hris/employee/CertificateStorageService.java
package com.compnet.hris.employee;

import com.compnet.hris.common.InvalidRequestException;
import com.compnet.hris.common.ResourceNotFoundException;
import com.compnet.hris.config.HrisProperties;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.Arrays;
import java.util.UUID;
import org.springframework.core.io.Resource;
import org.springframework.core.io.UrlResource;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

@Service
public class CertificateStorageService {
    private static final byte[] PDF = {'%', 'P', 'D', 'F', '-'};
    private static final byte[] JPEG = {(byte) 0xff, (byte) 0xd8, (byte) 0xff};
    private static final byte[] PNG = {(byte) 0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a};
    private final Path directory;
    private final long maxBytes;

    public CertificateStorageService(HrisProperties properties) {
        this.directory = properties.upload().directory().toAbsolutePath().normalize();
        this.maxBytes = properties.upload().maxBytes();
    }

    public StoredCertificate store(MultipartFile file) {
        if (file == null || file.isEmpty()) throw new InvalidRequestException("A certificate file is required");
        if (file.getSize() > maxBytes) throw new InvalidRequestException("Certificate exceeds the configured size limit");
        try {
            byte[] bytes = file.getBytes();
            FileType type = detect(bytes);
            if (type == null) throw new InvalidRequestException("Certificate must be a valid PDF, JPEG, or PNG file");
            Files.createDirectories(directory);
            String storedName = UUID.randomUUID() + "." + type.extension;
            Files.write(safePath(storedName), bytes, StandardOpenOption.CREATE_NEW);
            String original = safeOriginalName(file.getOriginalFilename());
            return new StoredCertificate(storedName, original, type.mimeType, bytes.length);
        } catch (IOException exception) {
            throw new IllegalStateException("Certificate storage failed", exception);
        }
    }

    public Resource open(String storedName) {
        try {
            Resource resource = new UrlResource(safePath(storedName).toUri());
            if (!resource.exists() || !resource.isReadable()) throw new ResourceNotFoundException("Certificate not found");
            return resource;
        } catch (IOException exception) {
            throw new ResourceNotFoundException("Certificate not found");
        }
    }

    public void remove(String storedName) {
        if (storedName == null) return;
        try { Files.deleteIfExists(safePath(storedName)); }
        catch (IOException exception) { throw new IllegalStateException("Certificate cleanup failed", exception); }
    }

    private Path safePath(String storedName) {
        Path result = directory.resolve(storedName).normalize();
        if (!result.startsWith(directory)) throw new ResourceNotFoundException("Certificate not found");
        return result;
    }

    private String safeOriginalName(String originalName) {
        if (originalName == null || originalName.isBlank()) return "certificate";
        String clean = Path.of(originalName.replace("\0", "").replace('\\', '/')).getFileName().toString();
        return clean.substring(0, Math.min(clean.length(), 255));
    }

    private FileType detect(byte[] bytes) {
        if (startsWith(bytes, PDF)) return new FileType("pdf", "application/pdf");
        if (startsWith(bytes, JPEG)) return new FileType("jpg", "image/jpeg");
        if (startsWith(bytes, PNG)) return new FileType("png", "image/png");
        return null;
    }

    private boolean startsWith(byte[] bytes, byte[] signature) {
        return bytes.length >= signature.length
                && Arrays.equals(Arrays.copyOf(bytes, signature.length), signature);
    }

    public record StoredCertificate(String storedName, String originalName, String mimeType, int size) {}
    private record FileType(String extension, String mimeType) {}
}
